logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb

```diff
@@ -6,130 +6,234 @@
-        [::File.exist?(PROC_SELF_CGROUP_FILE),
-         Dir.exist?(PROC_CGROUP_CPU_DIR),
-         Dir.exist?(PROC_CGROUP_CPUACCT_DIR)].all?
+      def cgroup_available?
+        # don't cache to ivar, in case the files are mounted after logstash starts??
+        @cgroup_file.exist? && @cpuacct_dir.exist? && @cpu_dir.exist?
       end
 
-      def control_groups
+      def controller_groups
         response = {}
-
-        read_proc_self_cgroup_lines.each do |line|
+        read_lines.each do |line|
           matches = CONTROL_GROUP_RE.match(line)
+          next if matches.nil?
           # multiples controls, same hierarchy
           controllers = matches[1].split(CONTROLLER_SEPARATOR_RE)
-          controllers.each_with_object(response) { |controller| response[controller] = matches[2] }
+          controllers.each_with_object(response) do |controller|
+            case controller
+            when @controller_cpu
+              response[controller] = CpuResource.new(@cpu_dir, @cpu_path_override, matches[2])
+            when @controller_cpuacct
+              response[controller] = CpuAcctResource.new(@cpuacct_dir, @cpuacct_path_override, matches[2])
+            else
+              response[controller] = UnhandledResource.new(Pathname.new("/sys/fs/cgroup").join(controller), NullOverride.new, matches[2])
+            end
+          end
         end
-
         response
       end
 
-      def read_first_line(path)
-        IO.readlines(path).first
+      def read_lines
+        IO.readlines(@cgroup_file)
       end
+    end
 
-      def cgroup_cpuacct_usage_nanos(control_group)
-        read_first_line(::File.join(PROC_CGROUP_CPUACCT_DIR, control_group, CPUACCT_USAGE_FILE)).to_i
+    module ControllerResource
+      attr_reader :base_path, :override, :offset_path
+      def initialize(base_path, override, offset_path)
+        @base_path = base_path
+        # override is needed here for the logging statements
+        # so we do the override here and not in the caller
+        @override = override
+        @offset_path = @override.override(offset_path)
       end
-
-      def cgroup_cpu_fs_period_micros(control_group)
-        read_first_line(::File.join(PROC_CGROUP_CPUACCT_DIR, control_group, CPU_FS_PERIOD_US_FILE)).to_i
+      def check_path(path, not_found_value)
+        if exist?(path)
+          yield
+        else
+          message = "File #{path} cannot be found, "
+          if override.nil?
+            message.concat("try providing an override '#{override.key}' in the Logstash JAVA_OPTS environment variable")
+          else
+            message.concat("even though the '#{override.key}' override is: '#{override.value}'")
+          end
+          logger.debug(message)
+          not_found_value
+        end
       end
-
-      def cgroup_cpu_fs_quota_micros(control_group)
-        read_first_line(::File.join(PROC_CGROUP_CPUACCT_DIR, control_group,  CPU_FS_QUOTA_US_FILE)).to_i
+      def exist?(path)
```

Comment:
that nearly every method in this module takes a `path` argument, and the path is a result of combining instance variables with something in an implementation's other method(s), leads me to believe we haven't found the most useful level of abstraction.